src/v/ssx/async_algorithm.h

```diff
@@ -347,4 +347,63 @@ async_for_each_counter(async_counter& counter, Container&& container, Fn f) {
       counter, std::begin(container), std::end(container), std::move(f));
 }
 
+/**
+ * @brief Call f until cond is false, yielding occasionally and accepting
+ * an externally provided counter for yield control.
+ *
+ * This is equivalent to while (cond()) f(); except that the computational
+ * loop yields every Traits::interval (default 100) iterations in order
+ * to avoid reactor stalls. The returned future resolves when cond() is false.
+ *
+ * The counter is taken by reference and must live at least until the
+ * returned future resolves: this usually trivial when the caller is a
+ * coroutine but may require some care when continuation style is used.
+ *
+ * The functions are taken by value.
+ *
+ * @param counter the counter object to use, may be reused across invocations
+ * @param cond the function to call to check the condition
+ * @param f the function to call on each element
+ * @return ss::future<> a future which resolves when !cond()
+ */
+template<
+  typename Traits = async_algo_traits,
+  std::predicate Cond,
+  std::invocable Fn>
+ss::future<> async_while_counter(async_counter& counter, Cond cond, Fn f) {
```

Comment:
It is probably worth mentioning whether `cond` and `f` can be or must be sync or async. I guess the answer is "must be sync" for both but worth mentioning and perhaps enforcing via concept. You could contrast with e.g., `do_until` in the "why/when do you want this" section.